File: packages/mcp/src/lib/toolFailureDiagnostics.ts
import { promises as fs } from "node:fs";
import path from "node:path";

type JsonRecord = Record<string, unknown>;

export type McpToolFailureEvent = {
  timestamp: string;
  tool: string;
  operation: string;
  input_path: string | null;
  resolved_path: string | null;
  error: {
    code: string | null;
    name: string | null;
    message: string;
  };
  cwd: string;
  vault_root: string | null;
  request_id: string | number | null;
  session_id: string | null;
  correlation_id: string | null;
};

export type McpToolFailureTypeSummary = {
  tool: string;
  error_code: string | null;
  error_name: string | null;
  count: number;
  first_timestamp: string;
  last_timestamp: string;
  sample_message: string;
};

export type McpToolFailureReport = {
  enabled: boolean;
  log_dir: string | null;
  log_path: string | null;
  scanned_events: number;
  matched_events: number;
  first_timestamp: string | null;
  last_timestamp: string | null;
  top_error_types: McpToolFailureTypeSummary[];
  recent_events: McpToolFailureEvent[];
};

export type LogMcpToolFailureOptions = {
  tool: string;
  operation?: string;
  args?: unknown;
  error: unknown;
  requestId?: unknown;
  sessionId?: unknown;
};

export type GetToolFailureReportOptions = {
  recentLimit: number;
  topErrorLimit: number;
  tool?: string;
};

export type McpToolFailureDiagnostics = {
  enabled: boolean;
  logDir: string | null;
  logPath: string | null;
  logToolFailure(options: LogMcpToolFailureOptions): Promise<void>;
  getToolFailureReport(options: GetToolFailureReportOptions): Promise<McpToolFailureReport>;
};

const LOG_FILE_NAME = "mcp-tool-failures.jsonl";

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function toNullableString(value: unknown): string | null {
  if (typeof value === "string") return value;
  if (typeof value === "number" && Number.isFinite(value)) return String(value);
  return null;
}

function toRequestId(value: unknown): string | number | null {
  if (typeof value === "string") return value;
  if (typeof value === "number" && Number.isFinite(value)) return value;
  return null;
}

function extractPrimaryInputPath(args: unknown): string | null {
  if (!isRecord(args)) return null;

  const pathLikeKeys = ["path", "input_path", "from_path", "to_path"] as const;
  for (const key of pathLikeKeys) {
    const value = args[key];
    if (typeof value === "string" && value.trim().length > 0) {
      return value;
    }
  }

  return null;
}

function normalizeError(error: unknown): {
  code: string | null;
  name: string | null;
  message: string;
} {
  if (error instanceof Error) {
    const rawCode = (error as NodeJS.ErrnoException).code;
    const code =
      typeof rawCode === "string" ? rawCode : typeof rawCode === "number" ? String(rawCode) : null;
    return {
      code,
      name: error.name || null,
      message: error.message || "Unknown error",
    };
  }

  if (isRecord(error)) {
    const code = toNullableString(error["code"]);
    const name = toNullableString(error["name"]);
    const message = toNullableString(error["message"]) ?? JSON.stringify(error);
    return { code, name, message };
  }

  return { code: null, name: null, message: String(error) };
}

function correlationIdFor(
  requestId: string | number | null,
  sessionId: string | null,
): string | null {
  if (requestId === null && sessionId === null) return null;
  const sid = sessionId ?? "no-session";
  const rid = requestId ?? "no-request";
  return `${sid}:${rid}`;
}

function toResolvedPath(vaultRoot: string | null, inputPath: string | null): string | null {
  if (!vaultRoot || !inputPath) return null;
  return path.resolve(vaultRoot, inputPath);
}

function parseEvent(raw: unknown): McpToolFailureEvent | null {
  if (!isRecord(raw)) return null;
  if (!isRecord(raw["error"])) return null;

  const timestamp = toNullableString(raw["timestamp"]);
  const tool = toNullableString(raw["tool"]);
  const operation = toNullableString(raw["operation"]);
  const cwd = toNullableString(raw["cwd"]);
  const message = toNullableString(raw["error"]["message"]);

  if (!timestamp || !tool || !operation || !cwd || !message) return null;

  return {
    timestamp,
    tool,
    operation,
    input_path: toNullableString(raw["input_path"]),
    resolved_path: toNullableString(raw["resolved_path"]),
    error: {
      code: toNullableString(raw["error"]["code"]),
      name: toNullableString(raw["error"]["name"]),
      message,
    },
    cwd,
    vault_root: toNullableString(raw["vault_root"]),
    request_id: toRequestId(raw["request_id"]),
    session_id: toNullableString(raw["session_id"]),
    correlation_id: toNullableString(raw["correlation_id"]),
  };
}

async function readEventsFromJsonl(logPath: string | null): Promise<McpToolFailureEvent[]> {
  if (!logPath) return [];

  let text: string;
  try {
    text = await fs.readFile(logPath, "utf8");
  } catch (error) {
    const code = error instanceof Error ? (error as NodeJS.ErrnoException).code : undefined;
    if (code === "ENOENT") return [];
    throw error;
  }

  const events: McpToolFailureEvent[] = [];
  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const parsed = JSON.parse(trimmed) as unknown;
      const event = parseEvent(parsed);
      if (event) events.push(event);
    } catch {
      // Skip malformed rows to keep diagnostics resilient.
    }
  }

  return events;
}

function summarizeTopErrors(
  events: McpToolFailureEvent[],
  topErrorLimit: number,
): McpToolFailureTypeSummary[] {
  type Bucket = {
    tool: string;
    error_code: string | null;
    error_name: string | null;
    count: number;
    first_timestamp: string;
    last_timestamp: string;
    sample_message: string;
  };

  const buckets = new Map<string, Bucket>();

  for (const event of events) {
    const bucketKey = `${event.tool}::${event.error.code ?? ""}::${event.error.name ?? ""}`;
    const existing = buckets.get(bucketKey);
    if (!existing) {
      buckets.set(bucketKey, {
        tool: event.tool,
        error_code: event.error.code,
        error_name: event.error.name,
        count: 1,
        first_timestamp: event.timestamp,
        last_timestamp: event.timestamp,
        sample_message: event.error.message,
      });
      continue;
    }

    existing.count += 1;
    if (event.timestamp < existing.first_timestamp) existing.first_timestamp = event.timestamp;
    if (event.timestamp > existing.last_timestamp) {
      existing.last_timestamp = event.timestamp;
      existing.sample_message = event.error.message;
    }
  }

  return [...buckets.values()]
    .sort((a, b) => b.count - a.count || b.last_timestamp.localeCompare(a.last_timestamp))
    .slice(0, topErrorLimit);
}

export function createMcpToolFailureDiagnostics(options: {
  vaultPath: string | undefined;
  cwd: string;
}): McpToolFailureDiagnostics {
  const vaultRoot = options.vaultPath ?? null;
  const enabled = Boolean(vaultRoot);
  const logDir = vaultRoot ? path.join(vaultRoot, ".ailss", "logs") : null;
  const logPath = logDir ? path.join(logDir, LOG_FILE_NAME) : null;

  let writeChain: Promise<void> = Promise.resolve();

  const enqueueWrite = async (line: string): Promise<void> => {
    if (!logDir || !logPath) return;

    writeChain = writeChain
      .then(async () => {
        await fs.mkdir(logDir, { recursive: true });
        await fs.appendFile(logPath, `${line}\n`, "utf8");
      })
      .catch(() => undefined);

    await writeChain;
  };

  const logToolFailure = async (logOptions: LogMcpToolFailureOptions): Promise<void> => {
    if (!enabled) return;

    const inputPath = extractPrimaryInputPath(logOptions.args);
    const requestId = toRequestId(logOptions.requestId);
    const sessionId = toNullableString(logOptions.sessionId);

    const event: McpToolFailureEvent = {
      timestamp: new Date().toISOString(),
      tool: logOptions.tool,
      operation: logOptions.operation ?? "tool_call",
      input_path: inputPath,
      resolved_path: toResolvedPath(vaultRoot, inputPath),
      error: normalizeError(logOptions.error),
      cwd: options.cwd,
      vault_root: vaultRoot,
      request_id: requestId,
      session_id: sessionId,
      correlation_id: correlationIdFor(requestId, sessionId),
    };

    try {
      await enqueueWrite(JSON.stringify(event));
    } catch {
      // Never fail a tool call because diagnostics logging failed.
    }
  };

  const getToolFailureReport = async (
    reportOptions: GetToolFailureReportOptions,
  ): Promise<McpToolFailureReport> => {
    const recentLimit = Math.max(1, Math.floor(reportOptions.recentLimit));
    const topErrorLimit = Math.max(1, Math.floor(reportOptions.topErrorLimit));
    const toolFilter = reportOptions.tool?.trim();

    const events = await readEventsFromJsonl(logPath);
    const filtered = toolFilter ? events.filter((event) => event.tool === toolFilter) : events;

    const firstTimestamp = filtered.length > 0 ? (filtered[0]?.timestamp ?? null) : null;
    const lastTimestamp =
      filtered.length > 0 ? (filtered[filtered.length - 1]?.timestamp ?? null) : null;

    return {
      enabled,
      log_dir: logDir,
      log_path: logPath,
      scanned_events: events.length,
      matched_events: filtered.length,
      first_timestamp: firstTimestamp,
      last_timestamp: lastTimestamp,
      top_error_types: summarizeTopErrors(filtered, topErrorLimit),
      recent_events: filtered.slice(-recentLimit).reverse(),
    };
  };

  return {
    enabled,
    logDir,
    logPath,
    logToolFailure,
    getToolFailureReport,
  };
}

File: apps/api/src/ailss_api/tool_failure_diagnostics.py
from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path

LOG_FILE_NAME = "mcp-tool-failures.jsonl"


@dataclass(frozen=True)
class ToolFailureEvent:
    timestamp: str
    tool: str
    operation: str
    input_path: str | None
    resolved_path: str | None
    error: dict[str, str | None]
    cwd: str
    vault_root: str | None
    request_id: str | int | None
    session_id: str | None
    correlation_id: str | None


@dataclass(frozen=True)
class ToolFailureTypeSummary:
    tool: str
    error_code: str | None
    error_name: str | None
    count: int
    first_timestamp: str
    last_timestamp: str
    sample_message: str


@dataclass(frozen=True)
class ToolFailureReport:
    enabled: bool
    log_dir: str | None
    log_path: str | None
    scanned_events: int
    matched_events: int
    first_timestamp: str | None
    last_timestamp: str | None
    top_error_types: list[ToolFailureTypeSummary]
    recent_events: list[ToolFailureEvent]


def _to_nullable_string(value: object) -> str | None:
    if isinstance(value, str):
        return value
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return str(value)
    return None


def _to_request_id(value: object) -> str | int | None:
    if isinstance(value, str):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    return None


def _extract_primary_input_path(args: Mapping[str, object] | None) -> str | None:
    if args is None:
        return None
    for key in ("path", "input_path", "from_path", "to_path"):
        value = args.get(key)
        if isinstance(value, str) and value.strip():
            return value
    return None


def _normalize_error(error: Exception) -> dict[str, str | None]:
    code = getattr(error, "code", None)
    return {
        "code": str(code) if isinstance(code, (str, int)) else None,
        "name": error.__class__.__name__,
        "message": str(error) or "Unknown error",
    }


def _correlation_id_for(
    request_id: str | int | None,
    session_id: str | None,
) -> str | None:
    if request_id is None and session_id is None:
        return None
    return f"{session_id or 'no-session'}:{request_id or 'no-request'}"


def _resolved_path(vault_root: Path | None, input_path: str | None) -> str | None:
    if vault_root is None or input_path is None:
        return None
    return str((vault_root / input_path).resolve())


class ToolFailureDiagnostics:
    def __init__(self, *, vault_path: Path | None, cwd: Path) -> None:
        self._vault_root = vault_path.resolve() if vault_path is not None else None
        self._cwd = cwd.resolve()
        self.enabled = self._vault_root is not None
        self.log_dir = self._vault_root / ".ailss" / "logs" if self._vault_root else None
        self.log_path = self.log_dir / LOG_FILE_NAME if self.log_dir else None

    def log_tool_failure(
        self,
        *,
        tool: str,
        args: Mapping[str, object] | None,
        error: Exception,
        request_id: object = None,
        session_id: object = None,
        operation: str = "tool_call",
    ) -> None:
        if not self.enabled or self.log_dir is None or self.log_path is None:
            return

        input_path = _extract_primary_input_path(args)
        event = ToolFailureEvent(
            timestamp=self._now_iso(),
            tool=tool,
            operation=operation,
            input_path=input_path,
            resolved_path=_resolved_path(self._vault_root, input_path),
            error=_normalize_error(error),
            cwd=str(self._cwd),
            vault_root=str(self._vault_root) if self._vault_root else None,
            request_id=_to_request_id(request_id),
            session_id=_to_nullable_string(session_id),
            correlation_id=_correlation_id_for(
                _to_request_id(request_id),
                _to_nullable_string(session_id),
            ),
        )

        self.log_dir.mkdir(parents=True, exist_ok=True)
        with self.log_path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(event.__dict__, ensure_ascii=True))
            handle.write("\n")

    def get_tool_failure_report(
        self,
        *,
        recent_limit: int,
        top_error_limit: int,
        tool: str | None = None,
    ) -> ToolFailureReport:
        if self.log_path is None:
            return ToolFailureReport(
                enabled=False,
                log_dir=None,
                log_path=None,
                scanned_events=0,
                matched_events=0,
                first_timestamp=None,
                last_timestamp=None,
                top_error_types=[],
                recent_events=[],
            )

        events = self._read_events()
        matched = [event for event in events if tool is None or event.tool == tool]
        recent_events = sorted(matched, key=lambda event: event.timestamp, reverse=True)[
            :recent_limit
        ]
        top_error_types = self._summarize_top_errors(matched, top_error_limit)

        timestamps = [event.timestamp for event in matched]
        return ToolFailureReport(
            enabled=self.enabled,
            log_dir=str(self.log_dir) if self.log_dir else None,
            log_path=str(self.log_path),
            scanned_events=len(events),
            matched_events=len(matched),
            first_timestamp=min(timestamps) if timestamps else None,
            last_timestamp=max(timestamps) if timestamps else None,
            top_error_types=top_error_types,
            recent_events=recent_events,
        )

    def _read_events(self) -> list[ToolFailureEvent]:
        if self.log_path is None or not self.log_path.exists():
            return []

        events: list[ToolFailureEvent] = []
        for raw_line in self.log_path.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line:
                continue
            try:
                parsed = json.loads(line)
            except ValueError:
                continue
            if not isinstance(parsed, dict):
                continue

            error = parsed.get("error")
            if not isinstance(error, dict):
                continue

            timestamp = parsed.get("timestamp")
            tool = parsed.get("tool")
            operation = parsed.get("operation")
            cwd = parsed.get("cwd")
            message = error.get("message")
            if not all(
                isinstance(value, str) for value in (timestamp, tool, operation, cwd, message)
            ):
                continue
            assert isinstance(timestamp, str)
            assert isinstance(tool, str)
            assert isinstance(operation, str)
            assert isinstance(cwd, str)
            assert isinstance(message, str)

            events.append(
                ToolFailureEvent(
                    timestamp=timestamp,
                    tool=tool,
                    operation=operation,
                    input_path=_to_nullable_string(parsed.get("input_path")),
                    resolved_path=_to_nullable_string(parsed.get("resolved_path")),
                    error={
                        "code": _to_nullable_string(error.get("code")),
                        "name": _to_nullable_string(error.get("name")),
                        "message": message,
                    },
                    cwd=cwd,
                    vault_root=_to_nullable_string(parsed.get("vault_root")),
                    request_id=_to_request_id(parsed.get("request_id")),
                    session_id=_to_nullable_string(parsed.get("session_id")),
                    correlation_id=_to_nullable_string(parsed.get("correlation_id")),
                )
            )
        return events

    def _summarize_top_errors(
        self,
        events: list[ToolFailureEvent],
        limit: int,
    ) -> list[ToolFailureTypeSummary]:
        buckets: dict[tuple[str, str | None, str | None], ToolFailureTypeSummary] = {}
        for event in events:
            key = (event.tool, event.error["code"], event.error["name"])
            existing = buckets.get(key)
            if existing is None:
                buckets[key] = ToolFailureTypeSummary(
                    tool=event.tool,
                    error_code=event.error["code"],
                    error_name=event.error["name"],
                    count=1,
                    first_timestamp=event.timestamp,
                    last_timestamp=event.timestamp,
                    sample_message=event.error["message"] or "Unknown error",
                )
                continue

            buckets[key] = ToolFailureTypeSummary(
                tool=existing.tool,
                error_code=existing.error_code,
                error_name=existing.error_name,
                count=existing.count + 1,
                first_timestamp=min(existing.first_timestamp, event.timestamp),
                last_timestamp=max(existing.last_timestamp, event.timestamp),
                sample_message=(
                    event.error["message"]
                    if event.timestamp >= existing.last_timestamp
                    else existing.sample_message
                )
                or "Unknown error",
            )

        return sorted(
            buckets.values(),
            key=lambda item: (-item.count, item.last_timestamp),
            reverse=False,
        )[:limit]

    def _now_iso(self) -> str:
        from datetime import datetime

        return datetime.now().isoformat(timespec="seconds")
